aws/s3_test.go

```diff
@@ -0,0 +1,291 @@
+		{
+			"Key":   "cloud-nuke-excluded",
+			"Value": "true",
+		},
+	})
+}
+
+func TestNukeEmptyS3Bucket(t *testing.T) {
+	s := setupNukeTests(t)
+
+	// Create test bucket
+	bucket := TestS3Bucket{
+		name: s.bucketName,
+	}
+
+	err := bucket.create(s.svc)
+	if err != nil {
+		assert.Failf(t, "Failed to create test bucket", errors.WithStackTrace(err).Error())
+	}
+
+	defer nukeAllS3Buckets(s.session, []*string{aws.String(s.bucketName)}, 1000)
+
+	// Nuke the test bucket
+	_, err = nukeAllS3Buckets(s.session, []*string{aws.String(s.bucketName)}, 1000)
+	if err != nil {
+		assert.Fail(t, errors.WithStackTrace(err).Error())
+	}
+
+	// Verify that - after nuking test bucket - it should not exist
+	bucketNamesPerRegion, err := getAllS3Buckets(s.session, time.Now().Add(1*time.Hour), s.bucketName)
+	if err != nil {
+		assert.Failf(t, "Failed to list S3 Buckets", errors.WithStackTrace(err).Error())
+	} else {
+		assert.NotContains(t, bucketNamesPerRegion[s.region], aws.String(s.bucketName))
+	}
+}
+
+func testNukeS3BucketWrapper(t *testing.T, isVersioned bool, objectCount int, objectBatchsize int) {
+	s := setupNukeTests(t)
+
+	// Create test bucket
+	bucket := TestS3Bucket{
+		name:        s.bucketName,
+		isVersioned: isVersioned,
+	}
+	err := bucket.create(s.svc)
+	if err != nil {
+		assert.Failf(t, "Failed to create test bucket", errors.WithStackTrace(err).Error())
+	}
+
+	defer nukeAllS3Buckets(s.session, []*string{aws.String(s.bucketName)}, 1000)
+
+	objectVersions := 1
+	if isVersioned {
+		objectVersions = 3
+	}
+
+	// Add two more versions of the same file
+	for i := 0; i < objectVersions; i++ {
+		for j := 0; j < objectCount; j++ {
+			fileName := fmt.Sprintf("l1/l2/l3/f%d.txt", j)
+			fileBody := fmt.Sprintf("%d-%d", i, j)
+			err := bucket.addObject(s, fileName, fileBody)
+			if err != nil {
+				assert.Failf(t, "Failed to add object to test bucket", errors.WithStackTrace(err).Error())
+			}
+		}
+	}
+
+	// Nuke the test bucket
+	_, err = nukeAllS3Buckets(s.session, []*string{aws.String(s.bucketName)}, objectBatchsize)
+	if err != nil {
+		assert.Fail(t, errors.WithStackTrace(err).Error())
+	}
+
+	// Verify that - after nuking test bucket - it should not exist
+	bucketNamesPerRegion, err := getAllS3Buckets(s.session, time.Now().Add(1*time.Hour), s.bucketName)
+	if err != nil {
+		assert.Failf(t, "Failed to list S3 Buckets", errors.WithStackTrace(err).Error())
+	} else {
+		assert.NotContains(t, bucketNamesPerRegion[s.region], aws.String(s.bucketName))
+	}
+}
+
+func TestNukeS3BucketWithoutVersioningAllObjects(t *testing.T) {
+	testNukeS3BucketWrapper(t, false, 10, 1000)
+}
+
+func TestNukeS3BucketWithoutVersioningBatchObjects(t *testing.T) {
+	testNukeS3BucketWrapper(t, false, 10, 2)
+}
+
+func TestNukeS3BucketWithVersioningAllObjects(t *testing.T) {
+	testNukeS3BucketWrapper(t, true, 10, 1000)
+}
+
+func TestNukeS3BucketWithVersioningBatchObjects(t *testing.T) {
+	testNukeS3BucketWrapper(t, true, 10, 3)
+}
```

Comment:
Thanks for adding these tests! One more test request: nuking a bucket where an object was deleted on a versioned bucket. This requires using `DeleteMarkers` to empty, which you aren't handling.

Reply:
Done

Reply:
Will update this thread once the above query is resolved.

Reply:
Done